File: lib/domain/entities/bundle.dart
import 'dart:convert';

enum BundleDestinationScope { direct, broadcast }

enum BundlePriority { low, normal, high, critical }

class Bundle {
  const Bundle({
    required this.bundleId,
    required this.type,
    required this.sourceNodeId,
    this.sourcePublicKey,
    this.destinationNodeId,
    this.destinationScope = BundleDestinationScope.direct,
    this.priority = BundlePriority.normal,
    this.ackForBundleId,
    this.payload,
    this.payloadReference,
    this.signature,
    this.appId = 'offlimu.chat',
    required this.createdAt,
    this.expiresAtOverride,
    required this.ttlSeconds,
    this.hopCount = 0,
    this.acknowledged = false,
    this.sentAt,
    this.failedAttempts = 0,
    this.lastError,
  });

  static const String typeChatMessage = 'chat_message';
  static const String typeFileShareMetadata = 'file_share_metadata';
  static const String typeFileShareChunk = 'file_share_chunk';
  static const String typeAck = 'ack';
  static const String typeSyncRejection = 'sync_rejection';

  final String bundleId;
  final String type;
  final String sourceNodeId;
  final String? sourcePublicKey;
  final String? destinationNodeId;
  final BundleDestinationScope destinationScope;
  final BundlePriority priority;
  final String? ackForBundleId;
  final String? payload;
  final String? payloadReference;
  final String? signature;
  final String appId;
  final DateTime createdAt;
  final DateTime? expiresAtOverride;
  final int ttlSeconds;
  final int hopCount;
  final bool acknowledged;
  final DateTime? sentAt;
  final int failedAttempts;
  final String? lastError;

  bool get isAck => type == typeAck;
  bool get isSyncRejection => type == typeSyncRejection;

  String get signaturePayload => jsonEncode(<String, Object?>{
    'bundleId': bundleId,
    'type': type,
    'sourceNodeId': sourceNodeId,
    'sourcePublicKey': sourcePublicKey,
    'destinationNodeId': destinationNodeId,
    'destinationScope': destinationScope.name,
    'priority': priority.name,
    'ackForBundleId': ackForBundleId,
    'payload': payload,
    'payloadReference': payloadReference,
    'appId': appId,
    'createdAtMs': createdAt.millisecondsSinceEpoch,
    'expiresAtMs': expiresAtOverride?.millisecondsSinceEpoch,
    'ttlSeconds': ttlSeconds,
    'hopCount': hopCount,
  });

  bool get isBroadcast =>
      destinationScope == BundleDestinationScope.broadcast ||
      destinationNodeId == null;
  bool get hasInlinePayload => (payload ?? '').isNotEmpty;
  bool get hasPayloadReference => (payloadReference ?? '').isNotEmpty;

  DateTime get expiresAt =>
      expiresAtOverride ?? createdAt.add(Duration(seconds: ttlSeconds));
  bool get isExpired => DateTime.now().isAfter(expiresAt);

  static BundleDestinationScope destinationScopeFromWire(String? value) {
    return switch (value) {
      'broadcast' => BundleDestinationScope.broadcast,
      _ => BundleDestinationScope.direct,
    };
  }

  static BundlePriority priorityFromWire(String? value) {
    return switch (value) {
      'low' => BundlePriority.low,
      'high' => BundlePriority.high,
      'critical' => BundlePriority.critical,
      _ => BundlePriority.normal,
    };
  }

  bool isForwardable({required int maxHopCount}) {
    return !isExpired && hopCount < maxHopCount;
  }

  Bundle copyWith({
    String? bundleId,
    String? type,
    String? sourceNodeId,
    String? sourcePublicKey,
    String? destinationNodeId,
    BundleDestinationScope? destinationScope,
    BundlePriority? priority,
    String? ackForBundleId,
    String? payload,
    String? payloadReference,
    String? signature,
    String? appId,
    DateTime? createdAt,
    DateTime? expiresAtOverride,
    int? ttlSeconds,
    int? hopCount,
    bool? acknowledged,
    DateTime? sentAt,
    int? failedAttempts,
    String? lastError,
  }) {
    return Bundle(
      bundleId: bundleId ?? this.bundleId,
      type: type ?? this.type,
      sourceNodeId: sourceNodeId ?? this.sourceNodeId,
      sourcePublicKey: sourcePublicKey ?? this.sourcePublicKey,
      destinationNodeId: destinationNodeId ?? this.destinationNodeId,
      destinationScope: destinationScope ?? this.destinationScope,
      priority: priority ?? this.priority,
      ackForBundleId: ackForBundleId ?? this.ackForBundleId,
      payload: payload ?? this.payload,
      payloadReference: payloadReference ?? this.payloadReference,
      signature: signature ?? this.signature,
      appId: appId ?? this.appId,
      createdAt: createdAt ?? this.createdAt,
      expiresAtOverride: expiresAtOverride ?? this.expiresAtOverride,
      ttlSeconds: ttlSeconds ?? this.ttlSeconds,
      hopCount: hopCount ?? this.hopCount,
      acknowledged: acknowledged ?? this.acknowledged,
      sentAt: sentAt ?? this.sentAt,
      failedAttempts: failedAttempts ?? this.failedAttempts,
      lastError: lastError ?? this.lastError,
    );
  }
}

File: server/src/types/bundle.ts
import { z } from 'zod';

export const bundleDestinationScopeSchema = z.enum(['direct', 'broadcast']);
export const bundlePrioritySchema = z.enum(['low', 'normal', 'high', 'critical']);

export const wireBundleSchema = z.object({
  bundleId: z.string().min(1),
  type: z.string().min(1),
  sourceNodeId: z.string().min(1),
  sourcePublicKey: z.string().nullable().optional(),
  destinationNodeId: z.string().nullable().optional(),
  destinationScope: bundleDestinationScopeSchema.default('direct'),
  priority: bundlePrioritySchema.default('normal'),
  ackForBundleId: z.string().nullable().optional(),
  payload: z.string().nullable().optional(),
  payloadRef: z.string().nullable().optional(),
  signature: z.string().nullable().optional(),
  appId: z.string().default('offlimu.chat'),
  createdAtMs: z.number().int(),
  expiresAtMs: z.number().int().nullable().optional(),
  ttlSeconds: z.number().int().positive(),
  hopCount: z.number().int().nonnegative().default(0),
  acknowledged: z.boolean().default(false),
  sentAtMs: z.number().int().nullable().optional(),
  failedAttempts: z.number().int().nonnegative().default(0),
  lastError: z.string().nullable().optional()
});

export const uploadRequestSchema = z.object({
  bundles: z.array(wireBundleSchema)
});

export const fetchQuerySchema = z.object({
  sinceMs: z.coerce.number().int().nonnegative().default(0)
});

export type WireBundle = z.infer<typeof wireBundleSchema>;

export type Bundle = Omit<WireBundle, 'payloadRef'> & {
  payloadReference?: string | null;
};

export function fromWireBundle(wire: WireBundle): Bundle {
  const { payloadRef, ...rest } = wire;
  return { ...rest, payloadReference: payloadRef ?? null };
}

export function toWireBundle(bundle: Bundle): WireBundle {
  const { payloadReference, ...rest } = bundle;
  return {
    ...rest,
    payloadRef: payloadReference ?? null
  };
}

export function isExpired(bundle: Bundle, nowMs = Date.now()): boolean {
  const expiresAtMs = bundle.expiresAtMs ?? bundle.createdAtMs + bundle.ttlSeconds * 1000;
  return nowMs > expiresAtMs;
}

export function signaturePayload(bundle: Bundle): string {
  return JSON.stringify({
    bundleId: bundle.bundleId,
    type: bundle.type,
    sourceNodeId: bundle.sourceNodeId,
    sourcePublicKey: bundle.sourcePublicKey ?? null,
    destinationNodeId: bundle.destinationNodeId ?? null,
    destinationScope: bundle.destinationScope,
    priority: bundle.priority,
    ackForBundleId: bundle.ackForBundleId ?? null,
    payload: bundle.payload ?? null,
    payloadReference: bundle.payloadReference ?? null,
    appId: bundle.appId,
    createdAtMs: bundle.createdAtMs,
    expiresAtMs: bundle.expiresAtMs ?? null,
    ttlSeconds: bundle.ttlSeconds
  });
}

export function legacySignaturePayload(bundle: Bundle): string {
  return JSON.stringify({
    bundleId: bundle.bundleId,
    type: bundle.type,
    sourceNodeId: bundle.sourceNodeId,
    sourcePublicKey: bundle.sourcePublicKey ?? null,
    destinationNodeId: bundle.destinationNodeId ?? null,
    destinationScope: bundle.destinationScope,
    priority: bundle.priority,
    ackForBundleId: bundle.ackForBundleId ?? null,
    payload: bundle.payload ?? null,
    payloadReference: bundle.payloadReference ?? null,
    appId: bundle.appId,
    createdAtMs: bundle.createdAtMs,
    expiresAtMs: bundle.expiresAtMs ?? null,
    ttlSeconds: bundle.ttlSeconds,
    hopCount: bundle.hopCount
  });
}

export function parseJsonPayload(bundle: Bundle): Record<string, unknown> | null {
  if (!bundle.payload) return null;
  try {
    const parsed = JSON.parse(bundle.payload);
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed)
      ? (parsed as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}
